core/src/main/scala/kafka/server/KafkaConfig.scala

```diff
@@ -559,7 +559,16 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
     controllerListenerNames.flatMap { name =>
       controllerAdvertisedListeners
         .find(endpoint => endpoint.listenerName.equals(ListenerName.normalised(name)))
-        .orElse(controllerListenersValue.find(endpoint => endpoint.listenerName.equals(ListenerName.normalised(name))))
+        .orElse(
+          // implicit advertised controller listener
```

Comment:
nit: how do you feel about making the comment more verbose to explain why this conversion is needed?

Reply:
> would it have made sense to remove the condition that effectiveAdvertisedControllerListeners cannot be 0.0.0.0

Hi @ahuang98, thanks for the review. From the [document](https://kafka.apache.org/documentation/#brokerconfigs_advertised.listeners), it says `advertised.listeners` can't have `0.0.0.0` host name. I think we still need to keep the check to avoid users define `advertised.listeners` with `0.0.0.0` explicitly. 

> have withWildcardHostnamesResolved account for the 0.0.0.0 case?

My change here to remove `0.0.0.0` from implicit advertised controller listeners make `withWildcardHostnamesResolved` work for them.

Reply:
That's a bit what my question was (why can't we relax that rule), is it because we can't/shouldn't change the documentation for 3.9 or it's actually incorrect to do so for some reason I'm missing